languages/rust/template/src/bin/pipe.rs

```diff
@@ -0,0 +1,165 @@
+// This file provides interop with the langserver
+//
+use algorithm;
+
+use base64;
+use serde::{Deserialize, Serialize};
+use serde_json;
+
+use algorithmia::error::{Error, ErrorKind, ResultExt};
+use algorithmia::prelude::{AlgoIo, EntryPoint};
+use serde_json::Value;
+use std::error::Error as StdError;
+use std::fs::OpenOptions;
+use std::io::{self, BufRead, Write};
+use std::process;
+
+const ALGOOUT: &'static str = "/tmp/algoout";
+
+#[derive(Deserialize)]
+struct Request {
+    data: Value,
+    content_type: String,
+}
+
+#[derive(Serialize)]
+struct AlgoSuccess {
+    result: Value,
+    metadata: RunnerMetadata,
+}
+
+#[derive(Serialize)]
+struct AlgoFailure {
+    error: RunnerError,
+}
+
+#[derive(Serialize)]
+struct RunnerMetadata {
+    content_type: String,
+}
+
+#[derive(Serialize)]
+struct RunnerError {
+    message: String,
+    error_type: &'static str,
+}
+
+impl AlgoSuccess {
+    fn new<S: Into<String>>(result: Value, content_type: S) -> AlgoSuccess {
+        AlgoSuccess {
+            result: result,
+            metadata: RunnerMetadata {
+                content_type: content_type.into(),
+            },
+        }
+    }
+}
+
+impl AlgoFailure {
+    fn new(err: &dyn StdError) -> AlgoFailure {
+        AlgoFailure {
+            error: RunnerError {
+                message: error_cause_chain(err),
+                error_type: "AlgorithmError",
+            },
+        }
+    }
+
+    fn system(err: &dyn StdError) -> AlgoFailure {
+        AlgoFailure {
+            error: RunnerError {
+                message: error_cause_chain(err),
+                error_type: "SystemError",
+            },
+        }
+    }
+}
+
+fn main() {
+    let mut algo = algorithm::Algo::default();
+    println!("PIPE_INIT_COMPLETE");
+    flush_std_pipes();
+
+    let stdin = io::stdin();
+    for line in stdin.lock().lines() {
+        let output_json = match line {
+            Ok(input) => {
+                let output = call_algorithm(&mut algo, input);
```

Comment:
Interesting, so here we're expecting that the downstream algorithm developer will figure out how to parse the arbitrary input object (of which the type is undefined in this instance? I think it's a string?), keeping auto parsing work out of the pipe object itself.

I like this approach honestly, as it keeps things consistent with some of our more successful languages (python, for example) - but I wonder if we'd ever run into any issues with this when attempting to use it for other languages..

Reply:
And I see that you're also not parsing the input json object, `{"metadata":{"content_type":"..."}, "data": "..."}` - are we assuming that the algorithm developer will parse that themselves?
https://github.com/algorithmiaio/langpacks/pull/66/files#diff-f879cfb8ab5d0c3b2f32a894ad0d4d9bR12
https://github.com/algorithmiaio/langpacks/pull/66/files#diff-20313b420f1e2e73ffd5133ef2c6a1bfR32
^ for example in my PR

Reply:
The `call_algorithm` function uses `serde_json`   to deserialize that json object into `struct Request`  and then it turns it into `enum AlgoIo` and then calls `Algo::apply` which is usually generated by the `#[entrypoint]` macro so that the algorithm author can work directly with the type they care about rather than explicitly handling all possit inputs. I admit, this would be a whole lot simpler if not supporting binary input/output.

Reply:
everything I wrote above is no longer true after the latest update. and the former pipe.rs no longer exists.